staging/src/k8s.io/client-go/util/certificate/certificate_manager.go

```diff
@@ -386,12 +398,15 @@ func (m *manager) rotateCerts() (bool, error) {
 		return false, m.updateServerError(err)
 	}
 
+	ctx, cancel := context.WithTimeout(context.Background(), certificateWaitTimeout)
+	defer cancel()
+
 	// Once we've successfully submitted a CSR for this template, record that we did so
-	m.setLastRequest(template)
+	m.setLastRequest(cancel, template)
 
 	// Wait for the certificate to be signed. This interface and internal timout
 	// is a remainder after the old design using raw watch wrapped with backoff.
-	crtPEM, err := csr.WaitForCertificate(client, req, certificateWaitTimeout)
+	crtPEM, err := csr.WaitForCertificate(ctx, client, req)
 	if err != nil {
 		utilruntime.HandleError(fmt.Errorf("Certificate request was not signed: %v", err))
 		return false, nil
```

Comment:
Is it worthwhile to check the err type from `WaitForCertificate` and return a non-nil err here? Otherwise we would re-run rotateCert under `wait.ExponentialBackoff(backoff, m.rotateCerts)` but the context has already been cancelled.

Reply:
Ah, we would just assign a new context in that case, but still doing the extra work at that point of cancelling the context up to 5 times right?

Reply:
as soon as rotateCerts returns an error, exponential backoff short-circuits:

https://github.com/kubernetes/kubernetes/blob/4d3d153210c0788b8e188c467d31061470d8c042/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L280-L285

Reply:
heh, I take it back... rotateCerts logs but returns nil there. I'll look at propagating that

Reply:
nevermind, I take back taking it back. the next time rotateCerts is called it will request the new template we want